### DISCUSSION TITLE
Is there any liburing version macro for backward compatibility

### DISCUSSION THREAD

**Discussion by akankshamahajan15** [axboe/liburing#601] [Q&A] [answered]
In our developement code, in some hosts we use older version of liburing  and some places we use newer one.
Currently we can't update to new version everywhere.
`io_uring_prep_cancel` takes  `user_data` as`void*` in  older and `__u64` in new. 
Is there any macro version that I can use so that my code can be compatible with both liburing version.

**Comment by YoSTEALTH** [comment]
looks like both new and old `user_data` (`addr`) is set to `__u64` `struct io_uring_sqe { __u64 addr;}`.

As a monkey-patch could you not write a wrapper function or cast `void *` to `__u64` `(unsigned long) user_data` vice versa?

ideally `io_uring_prep_cancel64` should have been created.

**Reply by akankshamahajan15** [reply]
With the new version it gives error that cannot convert from void* to __u64 if I use `io_uring_prep_cancel(sqe, (void*)user_data`. 
If I use` io_uring_prep_cancel(sqe, (unsigned long)user_data` then i get error with older liburing version. 

I have hacked it to use
```
 io_uring_prep_cancel(sqe, 0, 0),
sqe->addr = (unsigned long)user_data;
```

But its not a clean solution and wondering if there is any version check that I can use.
For older version use older API call
and for new version use new API call.

**Comment by YoSTEALTH** [comment]
@axboe Looks like
`static inline void io_uring_prep_cancel(struct io_uring_sqe *sqe, void *user_data, int flags)`
was implemented in Liburing version `<= 2.1` how about keeping that as is, and also implementing
`static inline void io_uring_prep_cancel64(struct io_uring_sqe *sqe, __u64 user_data, int flags)` for next release?

**Comment by axboe** [comment, marked as answer]
@YoSTEALTH Yeah that's not a bad idea, I'll take a look. It was an unfortunate change in retrospect, but it's not in a release yet at least.

**Reply by axboe** [reply]
https://git.kernel.dk/cgit/liburing/commit/?id=5e63866e719c47f13a0c43c90019fae912b328d8
